test/system/src/main/java/io/pravega/test/system/framework/services/docker/DockerBasedService.java

```diff
@@ -0,0 +1,191 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.test.system.framework.services.docker;
+
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.spotify.docker.client.DefaultDockerClient;
+import com.spotify.docker.client.DockerClient;
+import com.spotify.docker.client.exceptions.DockerException;
+import com.spotify.docker.client.messages.swarm.EndpointSpec;
+import com.spotify.docker.client.messages.swarm.PortConfig;
+import com.spotify.docker.client.messages.swarm.Service;
+import com.spotify.docker.client.messages.swarm.ServiceMode;
+import com.spotify.docker.client.messages.swarm.ServiceSpec;
+import com.spotify.docker.client.messages.swarm.TaskSpec;
+import com.spotify.docker.client.messages.swarm.TaskStatus;
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.Futures;
+import static io.pravega.test.system.framework.DockerRemoteSequential.DOCKER_CLIENT_PORT;
+import lombok.extern.slf4j.Slf4j;
+import java.net.URI;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import com.spotify.docker.client.messages.Network;
+
+@Slf4j
+public abstract class DockerBasedService implements io.pravega.test.system.framework.services.Service {
+
+    static final int ZKSERVICE_ZKPORT = 2181;
+    static final String IMAGE_PATH = System.getProperty("dockerImageRegistry");
+    static final String PRAVEGA_VERSION = System.getProperty("imageVersion");
+    static final String DOCKER_NETWORK = "docker-network";
+    final DockerClient dockerClient;
+    final String serviceName;
+    private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(3);
+
+    DockerBasedService(final String serviceName) {
+        this.dockerClient = DefaultDockerClient.builder().uri("http://" + System.getProperty("masterIP") + ":" + DOCKER_CLIENT_PORT).build();
+        this.serviceName = serviceName;
+    }
+
+    @Override
+    public String getID() {
+        Service.Criteria criteria = Service.Criteria.builder().serviceName(this.serviceName).build();
+        String serviceId = null;
+        try {
+            List<Service> serviceList = Exceptions.handleInterrupted(() -> dockerClient.listServices(criteria));
+            serviceId = serviceList.get(0).id();
+        } catch (DockerException e) {
+            log.error("Unable to get service id", e);
```

Comment:
we use getID() only in certain places, to stop the service and it's rarely called.  it should be fine, as we fail the test in case we are unable to stop/remove the service.